README
Acme-SList-Utilities Version 0.02

Schedule batch files for the SList suite of programs

SYNOPSIS

    use Acme::SList::Utilities qw(crdir sdate sduration commify);

    crdir '../dir1/dir2/dir3' or die "Error: Can't create directory because $!";

    print 'today is: ', sdate(time), "\n";

    my $from = time;
    sleep(4);
    my $to = time;
    print 'Time passed: ', sduration($from, $to), "\n";

    my $number = 1234567.89876;
    print 'number commified: ', commify($number), "\n";

INSTALLATION

    To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

    After installing, you can find documentation for this module with the
    perldoc command.

    perldoc Acme::SList::Utilities

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Acme-SList-Utilities

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Acme-SList-Utilities

    CPAN Ratings
        http://cpanratings.perl.org/d/Acme-SList-Utilities

    Search CPAN
        http://search.cpan.org/dist/Acme-SList-Utilities/


AUTHOR

    Klaus Eichner <klaus03@gmail.com>

COPYRIGHT AND LICENSE

    Copyright (C) 2009-2011 by Klaus Eichner

    All rights reserved. This program is free software; you can redistribute
    it and/or modify it under the terms of the artistic license 2.0,
    see http://www.opensource.org/licenses/artistic-license-2.0.php


Changes
Revision history for Acme::SList::Utilities

0.03 - 2014-09-20T19:29:17+01:00

  [DOCUMENTATION]

  - Add a .gitignore file

  - Changelog dates are better written as YYYY.MM.DD and probably ordered in
    the other, descending, direction
    Remark: I have decided to follow the ISO 8601 standard at
    http://www.w3.org/TR/NOTE-datetime

  - Just use a Makefile.PL and not a Build.PL the latter does not have any
    added value any more
  - In Makefile.PL ==> meta files link to repository

0.02 - 2011-11-23T12:00:00+01:00

  [DOCUMENTATION]

  - Add File::Slurp to the list of required modules
    in Makefile.PL    PREREQ_PM  => { 'File::Slurp' => 0 }
    in Build.PL       requires   => { 'File::Slurp' => 0 }

0.01 - 2011-11-07T12:00:00+01:00

  [ENHANCEMENTS]

  - First version, released on an unsuspecting world.



Makefile.PL
use 5.008;

use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Acme::SList::Utilities',
    VERSION_FROM      => 'lib/Acme/SList/Utilities.pm', # finds $VERSION
    PREREQ_PM         => { 'File::Slurp' => 0 }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Acme/SList/Utilities.pm', # retrieve abstract from module
       AUTHOR         => 'Klaus Eichner <klaus03@gmail.com>') : ()),
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
               type => 'git',
               url  => 'https://github.com/klaus03/Acme-SList-Utilities.git',
               web  => 'https://github.com/klaus03/Acme-SList-Utilities',
           }
        }})
        : ()
    ),
);


lib/Acme/SList/Utilities.pm
package Acme::SList::Utilities;

use strict;
use warnings;

use File::Copy;
use File::Slurp;

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT    = qw();
our @EXPORT_OK = qw(crdir sdate sduration commify target dircopy);
our $VERSION   = '0.03';

sub crdir {
    my ($path) = @_;

    my $dir = '';
    for my $elem (split m{[/\\]}xms, $path) {
        if ($elem =~ m{\A \s* \z}xms) {
            $! = 33; # Domain error
            return;
        }
        $dir .= $elem.'/';
        if ($elem ne '..' and !-d $dir) {
            mkdir $dir or return;
        }
    }
    return 1;
}

sub dircopy {
    my ($from, $to) = @_;

    $from =~ s{/+ \z}''xms;
    $to   =~ s{/+ \z}''xms;

    for my $file (read_dir $from) {
        copy "$from/$file", "$to/$file" or return;
    }

    return 1;
}

sub sdate {
    my ($stamp) = @_;

    my ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday) = localtime $stamp;
    return sprintf("%02d/%02d/%04d %02d:%02d:%02d",
      $mday, $mon + 1, $year + 1900, $hour, $min, $sec);
}

sub sduration { # calculate a duration ($_[0]...$_[1])
    my ($from, $to) = @_;

    my $tsec  = $to - $from;     my $sec  = $tsec % 60;
    my $tmin  = int($tsec / 60); my $min  = $tmin % 60;
    my $thour = int($tmin / 60); my $hour = $thour;

    my $dur = "$sec sec";
    $dur = "$min min ".$dur  unless $min  == 0;
    $dur = "$hour hrs ".$dur unless $hour == 0;
    return $dur;
}

sub commify {
    my ($number) = @_;

    1 while $number =~ s/^([-+]?\d+)(\d{3})/$1_$2/;
    $number =~ s/\./,/;
    return $number;
}

sub target {
    my ($Program) = @_;

    if (defined($ENV{'SL_Target'}) and $ENV{'SL_Target'} ne '') {
        my $sl_target = lc $ENV{'SL_Target'};
        if ($sl_target !~ m{\A [0-9a-z]{0,4} \z}xms) {
            die "Variable SL_Target: '$sl_target' contains non-alphanumeric characters or is more than 4 characters long";
        }
        return "SList-$Program-$sl_target";
    }
    return "SList-$Program";
}

1;

__END__

=head1 NAME

Acme::SList::Utilities - Various utilitiy-functions for the SList suite of programs

=head1 SYNOPSIS

    use Acme::SList::Utilities qw(crdir sdate sduration commify);

    crdir '../dir1/dir2/dir3' or die "Error: Can't create directory because $!";

    print 'today is: ', sdate(time), "\n";

    my $from = time;
    sleep(4);
    my $to = time;
    print 'Time passed: ', sduration($from, $to), "\n";

    my $number = 1234567.89876;
    print 'number commified: ', commify($number), "\n";

=head1 AUTHOR

Klaus Eichner <klaus03@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009-2011 by Klaus Eichner

All rights reserved. This program is free software; you can redistribute
it and/or modify it under the terms of the artistic license 2.0,
see http://www.opensource.org/licenses/artistic-license-2.0.php

=cut
